OpenRA.Mods.Common/Effects/Missile.cs

```diff
@@ -141,52 +200,106 @@ public void Tick(World world)
+				// horizontal rate of turn is applied.
+				WVec vDist;
+				if (info.CruiseAltitude.Length > 0 && dist.HorizontalLength * info.VerticalRateOfTurn > info.Speed.Length * ((vFacing + 0x40) & 0xFF))
+					vDist = new WVec(-(dist.Z - targetPosition.Z + info.CruiseAltitude.Length + terrainHeight), -dist.HorizontalLength, 0);
+				else
+					vDist = new WVec(-dist.Z, -dist.HorizontalLength, 0);
+
+				// Compute which direction the projectile should be facing
+				var desiredHFacing = OpenRA.Traits.Util.GetFacing(dist, hFacing);
+				var desiredVFacing = OpenRA.Traits.Util.GetFacing(vDist, vFacing);
+
+				// Check whether the homing mechanism is jammed
+				var jammed = info.Jammable && world.ActorsWithTrait<JamsMissiles>().Any(JammedBy);
+				if (jammed)
+				{
+					desiredHFacing = hFacing + world.SharedRandom.Next(-20, 21);
+					desiredVFacing = vFacing + world.SharedRandom.Next(-20, 21);
+				}
+				else if (!args.GuidedTarget.IsValidFor(args.SourceActor))
+					desiredHFacing = hFacing;
+
+				// Compute new direction the projectile will be facing
+				hFacing = OpenRA.Traits.Util.TickFacing(hFacing, desiredHFacing, info.RateOfTurn);
+				vFacing = OpenRA.Traits.Util.TickFacing(vFacing, desiredVFacing, info.VerticalRateOfTurn);
+
+				// Compute the projectile's guided displacement
+				move = new WVec(0, -1024, 0)
+					.Rotate(new WRot(WAngle.FromFacing(vFacing), WAngle.Zero, WAngle.Zero))
+					.Rotate(new WRot(WAngle.Zero, WAngle.Zero, WAngle.FromFacing(hFacing)))
+					* info.Speed.Length / 1024;
 			}
-			else if (!args.GuidedTarget.IsValidFor(args.SourceActor))
-				desiredFacing = facing;
-
-			facing = OpenRA.Traits.Util.TickFacing(facing, desiredFacing, info.RateOfTurn);
-			var move = new WVec(0, -1024, 0).Rotate(WRot.FromFacing(facing)) * info.Speed.Length / 1024;
-
-			if (pos.Z != desiredAltitude)
+			else
 			{
-				var delta = move.HorizontalLength * info.MaximumPitch.Tan() / 1024;
-				var dz = (targetPosition.Z - pos.Z).Clamp(-delta, delta);
-				move += new WVec(0, 0, dz);
+				// Compute the projectile's freefall displacement
+				move = velocity + gravity / 2;
+				velocity += gravity;
 			}
 
+			// Move the projectile
 			pos += move;
 
-			if (!string.IsNullOrEmpty(info.Trail) && --ticksToNextSmoke < 0)
+			// Create the smoke trail effect
+			if (!string.IsNullOrEmpty(info.Trail) && --ticksToNextSmoke < 0 && (activated || info.SmokeWhenDeactivated))
 			{
-				world.AddFrameEndTask(w => w.Add(new Smoke(w, pos - 3 * move / 2, info.Trail, trailPalette, info.Sequence)));
+				world.AddFrameEndTask(w => w.Add(new Smoke(w, pos - 3 * move / 2, info.Trail, smokePalette, info.SmokeSequence)));
 				ticksToNextSmoke = info.TrailInterval;
 			}
 
 			if (info.ContrailLength > 0)
 				contrail.Update(pos);
 
-			var cell = world.Map.CellContaining(pos);
-
-			var shouldExplode = (pos.Z < 0) // Hit the ground
+			var shouldExplode = (pos.Z <= terrainHeight) // Hit the ground
```

Comment:
It will be steppier than it needs to be right now because terrain height doesn't yet account for slopes (I have code that fixes this, but i'm waiting for the other heightmap stuff to be integrated first).  Don't worry *too* much about TS right now - as long as you don't do anything that we know is incompatible, then we can build on it in future prs.